### PR TITLE
feat(models): add qwen3.5-35b-a3b-fp8 and qwen3.5-397b-a17b-fp8 model configs and HuggingFace mapping

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Adds FP8 weight-only model configurations, dynamic dequantization layer wiring, and Hugging Face checkpoint conversion mappings for **Qwen 3.5 35B** (`Qwen/Qwen3.5-35B-A3B-FP8`) and **Qwen 3.5 397B** (`Qwen/Qwen3.5-397B-A17B-FP8`).

> [!NOTE]
> The 35B variant is validated end-to-end (checkpoint conversion, forward logit parity, autoregressive decoding, and distributed training). The **397B variant is config-only**: it shares the identical architecture family, parameter mapping, and dynamic dequantization engine, and is validated via `configs_test.py` smoke tests, but has **not** been validated end-to-end at scale on hardware. Reviewers should treat 397B as an enablement configuration.

### Summary of Changes:
1. **Model Configurations (`src/maxtext/configs/models/`)**:
   - Added `qwen3.5-35b-a3b-fp8.yml` and `qwen3.5-397b-a17b-fp8.yml` configuring FP8 weight-only storage (`weight_dtype: "float8_e4m3fn"`, `weight_block_size: 128`, compute `dtype: "bfloat16"`).
   - Set `float32_gate_logits: True` for numerical stability in MoE routing.
   - Declaratively specified `unquantized_modules` matching official Hugging Face quantization configs to preserve high-precision (`bfloat16`) for embeddings, norms, routing gates (`gate`, `shared_expert_gate`), and linear attention projections (`conv1d`, `in_proj_ba`, `A_log`, `dt_bias`).
   - Both configs inherit `sparse_matmul: True` from `base.yml` (Megablox sparse matmul) for production training and inference.
2. **Model Definition & Dequantization Wiring (`src/maxtext/models/qwen3.py`, `qwen3_5.py`)**:
   - Passed `weight_block_size` to linear attention projections (`in_proj_qkvz`, `out_proj`) in `Qwen3NextGatedDeltaNet` to enable block-scaled dequantization.
   - Wired `get_weight_dtype(cfg, ...)` across `Qwen3NextGatedDeltaNet`, `Qwen3NextSparseMoeBlock`, and `Qwen3_5DecoderLayer` to respect `unquantized_modules`.
   - Updated weight initialization (`conv_kernel_init` and `a_log_init`) to sample in `float32` when target dtypes are sub-byte / FP8 types, preventing truncation or numerical instability during initialization.
3. **Precision & Normalization Fixes (`src/maxtext/layers/normalizations.py`)**:
   - Updated `Qwen3NextRMSNorm` to compute `(1.0 + scale)` in `float32` prior to casting to `bfloat16`, eliminating mantissa truncation on small scale offsets ($\sim 10^{-3}$).
4. **Parameter Mapping & 2D Block Scale Hooks (`src/maxtext/checkpoint_conversion/utils/param_mapping.py`)**:
   - Extended `QWEN3_5_MAXTEXT_TO_HF_PARAM_MAPPING` to map companion `kernel_scale` parameters for self-attention (Q, K, V, Out), linear attention (`in_proj_qkvz`, `out_proj`), MLP shared expert, and MoE routed experts across scanned and unscanned modes.
   - Added 2D block scale transposition and composite scale hook functions (`reshape_scale`, `split_qkvz_scale`, `split_ba_scale`, and `process_wi_0_wi_1_scale`) to properly split and concatenate compound scale tensors.
   - Registered `qwen3.5-35b-a3b-fp8`, `qwen3.5-35b-fp8`, and `qwen3.5-397b-a17b-fp8` in `PARAM_MAPPING` and `HOOK_FNS`.
5. **Hugging Face Model Registry & Types**:
   - Registered FP8 model identifiers in `HF_IDS` (`src/maxtext/utils/globals.py`), `HF_MODEL_CONFIGS`, `HF_SHAPE` (`src/maxtext/checkpoint_conversion/utils/`), `_MODEL_TO_BLOCKS` (`src/maxtext/multimodal/processor.py`), `src/maxtext/layers/nnx_decoders.py`, and `ModelName` (`src/maxtext/configs/types.py`).
6. **Unit Tests & Test Registries**:
   - Registered `qwen3.5-35b-a3b.yml`, `qwen3.5-35b-a3b-fp8.yml`, and `qwen3.5-397b-a17b-fp8.yml` in `tests/unit/configs_test.py` (`QWEN_CONFIGS`).

# Tests

### Unit Tests
- Unit tests for all Qwen model configurations:
  ```bash
  PYTHONPATH=src pytest tests/unit/configs_test.py -k "qwen" -v
  ```
  Result: 14 passed, 0 failed.
- Full configuration suite:
  ```bash
  PYTHONPATH=src pytest tests/unit/configs_test.py -v
  ```
  Result: 79 passed, 0 failed.
- Pyconfig test suite:
  ```bash
  PYTHONPATH=src pytest tests/unit/pyconfig_test.py -v
  ```
  Result: 29 passed, 0 failed.

### End-to-End Validation (Qwen 3.5 35B FP8)
- **Forward Pass Logit Parity vs Hugging Face Golden** ([Run Log](https://paste.googleplex.com/4561960650276864)):
  - Validated on TPU v4-8 against official Hugging Face reference outputs (`tests/utils/forward_pass_logit_checker.py`).
  - Achieved **max KL divergence of 0.00143** (well below the `< 0.01` parity threshold and `< 0.2` assertion limit), with **10/10 top-token overlap** and **1.0 Jaccard similarity**.
  - *Note on `sparse_matmul`*: The test harness explicitly sets `sparse_matmul=false` to use the dense reference MoE path during parity evaluation, bypassing Megablox sparse matmul numerical divergence (documented in `qwen3_next_vs_reference_test.py:L670-672`) to strictly isolate FP8 dequantization parity.
- **Autoregressive Decoding Verification** ([Run Log](https://paste.googleplex.com/6526063589588992)):
  - Executed on TPU v4-8 with exit code 0.
  - Successfully generated 63 tokens at **~4.3 tok/s (234.9 ms/step)** with coherent, fluent text generation on standard attention benchmark prompts.
- **Distributed Training Verification** ([Run Log](https://paste.googleplex.com/5924369256284160)):
  - Executed 10 training steps (steps 0–9) on a **TPU v5p-128 slice (64 chips, FSDP=16, EP=4)** with exit code 0 using the `c4` dataset, dequantizing from the FP8 checkpoint on load.
  - Demonstrated stable loss progression (**`lm_loss: 2.725`** at step 9) and verified asynchronous Orbax checkpoint saving at step 9.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py [hunk 1]
```diff
@@ -872,6 +872,10 @@ def QWEN3_5_MAXTEXT_TO_HF_PARAM_MAPPING(config, maxtext_config, scan_layers=Fals
   """
   num_main_layers = config["text_config"]["num_hidden_layers"]
   layer_cycle_interval = maxtext_config.inhomogeneous_layer_cycle_interval
+  is_quantized = getattr(maxtext_config, "weight_dtype", None) in (
+      "float8_e4m3fn",
+      "float8_e5m2",
+  )
```
Comment (snehalv2002) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py]: qwen fp8 only uses F8_E4M3. remove extraneous cases.

File: src/maxtext/configs/models/qwen3.5-35b-a3b-fp8.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+# MoE Specific Parameters
+# Set base_mlp_dim to match base_moe_mlp_dim to pass validation for fully MoE models.
+base_mlp_dim: 512
+base_moe_mlp_dim: 512
+num_experts: 256
+shared_experts: 1
+num_experts_per_tok: 8
+norm_topk_prob: True
+float32_gate_logits: True
+sparse_matmul: False
```
Comment (Shuwen-Fang) [inline src/maxtext/configs/models/qwen3.5-35b-a3b-fp8.yml]: why is sparse matmul false?

Comment (snehalv2002) [reply src/maxtext/configs/models/qwen3.5-35b-a3b-fp8.yml]: need to remove i used it during testing to get forward logit test to pass. Reason i added it was a comment in `tests/unit/qwen3_next_vs_reference_test.py:L670-672`: `Force the test to use the 'dense_matmul' path in the MoE layer, as the 'sparse_matmul' path was found to be numerically incorrect compared to the reference.` This is not specific to our quantization changes.

**Review by gobbleturk — APPROVED** [review]

**Review by Shuwen-Fang — APPROVED** [review]
